uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h

```diff
@@ -0,0 +1,795 @@
+    static bool Lift(const Type& aLowered, Type& aLifted) {                 \
+      aLifted = aLowered;                                                   \
+      return true;                                                          \
+    }                                                                       \
+    static Type Lower(const Type& aLifted) { return aLifted; }              \
+  }
+
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(uint8_t, ReadUInt8, WriteUInt8);
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(int8_t, ReadInt8, WriteInt8);
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(uint16_t, ReadUInt16, WriteUInt16);
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(int16_t, ReadInt16, WriteInt16);
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(uint32_t, ReadUInt32, WriteUInt32);
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(int32_t, ReadInt32, WriteInt32);
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(uint64_t, ReadUInt64, WriteUInt64);
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(int64_t, ReadInt64, WriteInt64);
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(float, ReadFloat, WriteFloat);
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(double, ReadDouble, WriteDouble);
+
+/// Booleans are passed as unsigned integers over the FFI, because JNA doesn't
+/// handle `bool`s well.
+
+template <>
+struct Serializable<bool> {
+  static size_t Size(const bool& aValue) { return 1; }
+  static bool ReadFrom(Reader& aReader, bool& aValue) {
+    aValue = aReader.ReadUInt8() != 0;
+    return true;
+  }
+  static void WriteInto(Writer& aWriter, const bool& aValue) {
+    aWriter.WriteUInt8(aValue ? 1 : 0);
+  }
+};
+
+template <>
+struct ViaFfi<bool, uint8_t> {
+  static bool Lift(const uint8_t& aLowered, bool& aLifted) {
+    aLifted = aLowered != 0;
+    return true;
+  }
+  static uint8_t Lower(const bool& aLifted) { return aLifted ? 1 : 0; }
+};
+
+/// Strings are length-prefixed and UTF-8 encoded when serialized
+/// into byte buffers, and are passed as null-terminated, UTF-8
+/// encoded `char*` pointers over the FFI.
+///
+/// Gecko has two string types: `nsCString` for "narrow" strings, and `nsString`
+/// for "wide" strings. `nsCString`s don't have a fixed encoding: these can be
+/// ASCII, Latin-1, or UTF-8. `nsString`s are always UTF-16. JS prefers
+/// `nsString` (UTF-16; also called `DOMString` in WebIDL); `nsCString`s
+/// (`ByteString` in WebIDL) are pretty uncommon.
+///
+/// `nsCString`s can be passed to Rust directly, and copied byte-for-byte into
+/// buffers. The UniFFI scaffolding code will ensure they're valid UTF-8. But
+/// `nsString`s must be converted to UTF-8 first.
+
+template <>
+struct Serializable<nsACString> {
+  static size_t Size(const nsACString& aValue) {
+    CheckedInt<size_t> size(aValue.Length());
+    size += sizeof(uint32_t);  // For the length prefix.
+    MOZ_RELEASE_ASSERT(size.isValid());
+    return size.value();
+  }
+
+  static bool ReadFrom(Reader& aReader, nsACString& aValue) {
+    aValue.Truncate();
```

Comment:
Just in case the out param has garbage in it.

Reply:
Again: put this in a comment :)